File: src/main/java/ru/randomsoftware/java2d/rules/NumericRule.java
package ru.randomsoftware.java2d.rules;

public class NumericRule implements Rule<Boolean> {

    private final boolean[] base;

    public NumericRule(int base) {
        if (base < 0 || base > 255) {
            throw new IllegalArgumentException("Base must be between 0 and 255");
        }
        this.base = intToBooleanArray(base);
    }

    @Override
    public Boolean calculate(Boolean a, Boolean b, Boolean c) {
        if (a && b && c) {
            return base[0];
        } else if (a && b) {
            return base[1];
        } else if (a && c) {
            return base[2];
        } else if (a) {
            return base[3];
        } else if (b && c) {
            return base[4];
        } else if (b) {
            return base[5];
        } else if (c) {
            return base[6];
        } else {
            return base[7];
        }
    }

    private static boolean[] intToBooleanArray(int number) {
        final boolean[] ret = new boolean[8];
        for (int i = 0; i < 8; i++) {
            ret[7 - i] = (1 << i & number) != 0;
        }
        return ret;
    }
}

File: src/main/kotlin/ru/randomsoftware/java2d/rules/NumericRule.kt
package ru.randomsoftware.java2d.rules

class NumericRule(base: Int): Rule<Boolean> {

    private val bs: List<Boolean>

    init {
        if (base < 0 || base > 255) {
            throw IllegalArgumentException("Base must be between 0 and 255")
        }
        bs = intToBooleanList(base)
    }

    override fun calculate(a: Boolean, b: Boolean, c: Boolean): Boolean {
        if (a && b && c) {
            return bs[0]
        } else if (a && b) {
            return bs[1]
        } else if (a && c) {
            return bs[2]
        } else if (a) {
            return bs[3]
        } else if (b && c) {
            return bs[4]
        } else if (b) {
            return bs[5]
        } else if (c) {
            return bs[6]
        } else {
            return bs[7]
        }
    }

    companion object {
        fun intToBooleanList(number: Int): List<Boolean> {
            return (7 downTo 0).map { 1 shl it and number != 0}
        }
    }
}